### ISSUE TITLE
[Feature Request] - Add `EnsembleKNeighborsClassifier`

### ISSUE THREAD

**Issue by jolespin** [jolespin/skclust#40]
# Implementation Plan: EnsembleKNeighborsClassifier

## Overview

Build an `EnsembleKNeighborsClassifier` class that ensembles KNN predictions across multiple k values using a shared similarity matrix. The class must be fully compatible with scikit-learn's estimator API.

## Class Signature

```python
class EnsembleKNeighborsClassifier(BaseEstimator, ClassifierMixin):
    def __init__(self, k_values, metric="cosine", weights="distance"):
```

## Parameters

- `k_values`: list of int — the k values to ensemble over (e.g., `[3, 5, 7, 11]`). No defaults; must be provided.
- `metric`: str — only `"cosine"` is currently supported. Default `"cosine"`. Raise `NotImplementedError` for any other value.
- `weights`: str — how to weight neighbors within each k. `"distance"` (default) weights neighbors by their cosine similarity. `"uniform"` weights all neighbors equally. Only these two are supported; raise `ValueError` for anything else.

## Validation (on fit or predict)

- `k_values` must be a list/array of positive integers with no duplicates.
- All k values must be ≤ `n_samples_train`. Raise `ValueError` otherwise.
- Input `X` must be 2D.

### `fit(X, y)`

- Store `X` as `self.X_train_` (reference embeddings, assumed L2-normalized).
- Store `y` as `self.y_train_`.
- Derive and store `self.classes_` via `np.unique(y)`.
- Build a label-to-index mapping for constructing probability vectors.
- Return `self`.

### `predict_proba_per_k(X)`

This is the core method. Everything else wraps it.

1. Compute similarity matrix: `S = X @ self.X_train_.T` — shape `(n_test, n_train)`.
2. Argsort each row in descending order once: `neighbor_indices = np.argsort(-S, axis=1)`.
3. Sort the similarity values in the same order for potential future use as weights.
4. For each k in `self.k_values`:
   - Slice the top-k neighbor indices: `neighbor_indices[:, :k]`.
   - Look up training labels for those neighbors.
   - Compute class probability vector per sample. If `weights="uniform"`, use fraction of neighbors per class. If `weights="distance"`, weight each neighbor's vote by its cosine similarity, then normalize to sum to 1.
5. Return a numpy array of shape `(n_samples, n_classes, n_k_values)`.

The ordering of the classes axis must correspond to `self.classes_`.

### `predict_proba(X)`

- Call `predict_proba_per_k(X)`.
- Return the mean over the k axis → shape `(n_samples, n_classes)`.

### `predict(X)`

- Call `predict_proba(X)`.
- Return `self.classes_[np.argmax(proba, axis=1)]`.

## sklearn Compatibility

- Inherit from `BaseEstimator` and `ClassifierMixin`.
- All constructor arguments must be stored as attributes with the same name (sklearn's `get_params` contract).
- Fitted attributes use trailing underscore convention (`X_train_`, `y_train_`, `classes_`).
- `predict_proba` and `predict` should call `check_is_fitted(self)` before executing.
- Use `sklearn.utils.validation.check_X_y` in `fit` and `check_array` in predict methods.

## Implementation Constraints

- **CPU/numpy only.** No torch, no GPU.
- Similarity computation uses `@` (matmul). Input embeddings are assumed L2-normalized so dot product = cosine similarity.
- The argsort is the most expensive operation and must happen exactly once per predict call, not once per k.
- Do not use `sklearn.neighbors.KNeighborsClassifier` internally.
## Coding Conventions

- Use `loguru` for logging (e.g., log k_values and training set size on fit).
- Multi-object imports on separate lines:
  ```python
  from sklearn.base import (
      BaseEstimator,
      ClassifierMixin,
  )
  ```

**Comment by jolespin** [comment]
https://github.com/jolespin/skclust/pull/41
